File: dotnet/Allmark/Render/ImageRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ImageRenderer
{
	public static Renderer Create()
	{
		return new Renderer
		{
			Name = "image",
			Render = Render,
		};
	}

	public static void Render(MarkdownNode node, RendererState state, bool? first = null, bool? last = null, bool? decode = true)
	{
		RenderUtils.StartNewLine(node, state);
		var alt = GetChildText(node);
		var title = !string.IsNullOrEmpty(node.Title) ? $" title=\"{node.Title}\"" : "";
		state.Output.Append($"<img src=\"{node.Info}\" alt=\"{alt}\"{title} />");
		RenderUtils.EndNewLine(node, state);
	}

	static string GetChildText(MarkdownNode node) {
		var text = "";
		if (node.Children != null) {
			foreach (var child in node.Children) {
				if (child.Type == "text") {
					text += child.Markup;
				} else {
					text += GetChildText(child);
				}
			}
		}
		return text;
	}
}

File: web/src/render/imageRenderer.ts
import type MarkdownNode from "../types/MarkdownNode";
import type Renderer from "../types/Renderer";
import type RendererState from "../types/RendererState";
import { forEachChild } from "../utils/nodeUtils";
import { endNewLine, startNewLine } from "./renderUtils";

const renderer: Renderer = {
	name: "image",
	render,
};
export default renderer;

function render(node: MarkdownNode, state: RendererState): void {
	startNewLine(node, state);
	let alt = getChildText(node);
	let title = node.title ? ` title="${node.title}"` : "";
	state.output += `<img src="${node.info}" alt="${alt}"${title} />`;
	endNewLine(node, state);
}

function getChildText(node: MarkdownNode): string {
	let text = "";
	forEachChild(node, (child) => {
		if (child.type === "text") {
			text += child.content;
		} else {
			text += getChildText(child);
		}
	});
	return text;
}
